text/204-golang-bindings.md

```diff
@@ -0,0 +1,669 @@
+    Cluster()                ICluster
+    DesiredCount()           *int
+    ServiceName()            *string
+    MaxHealthyPercent()      *int
+    MinHealthyPercent()      *int
+    HealthCheckGracePeriod() *Duration
+    CloudMapOptions()        CloudMapOptions
+    PropagateTags()          PropagatedTagSource
+    EnableECSManagedTags()   *bool
+    DeploymentController()   DeploymentController
+}
+
+type BaseServicePropsIface interface {
+    BaseServiceOptionsIface  // embeddeded interface
+    LaunchType()             LaunchType
+}
+
+type BaseServiceOptions struct {
+    cluster                ICluster
+    desiredCount           *int
+    serviceName            *string
+    maxHealthyPercent      *int
+    minHealthyPercent      *int
+    healthCheckGracePeriod *Duration
+    cloudMapOptions        CloudMapOptions
+    propagateTags          PropagatedTagSource
+    enableECSManagedTags   *bool
+    deploymentController   DeploymentController
+}
+
+type BaseServiceProps struct {
+    BaseServiceOptions  // embedded (anonymous) field
+    launchType          LaunchType
+}
+
+func (o BaseServiceOptions) Cluster() ICluster { return o.cluster }
+func (o BaseServiceOptions) DesiredCount() int { return o.desiredCount }
+
+// ... etc
+
+```
+
+This allows the embedding interface (here, `BaseServicePropsIFace`) to "inherit" all the methods defined in the embedded interface
+(`BaseServiceOptionsIFace`). Similarly, the embedding struct (`BaseServiceProps`) would have access to all the properties of the embedded struct
+(`BaseServiceOptions`) through having it as an anonymous field (See: [Go playground example](https://play.golang.org/p/XRpTH3LN6bx)).
+
+```go
+props := BaseServiceProps{
+    BaseServiceOptions{
+      serviceName: "myService",
+      ...
+    },
+    launchType: "EC2",
+}
+
+props.serviceName  // => "myService"
+```
+
+### Notes/Concerns
+
+* When fields in the Go struct are exported, the struct cannot have a method with the same name (e.g. exported field `Bar` , and method `Bar()`.
+   Using a `Bar()` getter on a private field bar is most consistent with idiomatic Go;  however, this may make code generation more difficult since it
+would potentially necessitate converting public properties on the source (JSII) class into an interface implemented by the custom struct (Go “class”).
+The *alternative* to this is to prefix the methods in the generated Go interface with Get, e.g. `GetBar()` to avoid the name collision.
+* Like the AWS Go SDK, we can use pointers to primitive types to simulate that fields are optional (i.e. allow null values rather than default “empty”
+  values for each type, which are not nullable)
+* Generated Go interfaces corresponding to a datatype interface would need a suffix, e.g. Iface,  in order to disambiguate it from the struct. This is
+  a bit verbose, and it may be worth considering switching the naming (i.e. adding a suffix to the struct instead) if the interface name is what will
```

Comment:
Not if they're supplying the struct values. Calling functions (building structs) is going to be more common than defining new structs.

Or can structs be anonymous?

Reply:
What do you mean? Interfaces in go do not supply values.

Structs can be anonymous but it's kind of clunky (see: https://play.golang.org/p/Tia8T67nMTi) and usually neater to wrap them up as a custom type.

Reply:
Right. I was responding to:

> if the interface name is what will primarily be used by the customer.

So:

* You type the *interface name* if you're **writing** a function that takes some data.
* You type the *struct name* if you're **calling** a function that takes some data.

The second one is likely to be more common, so that's the case we should optimize for. That means the best naming convention is probably the one where the struct has the type name, and the interface has `<typename>Iface` (or whatever).